cudaextender/include/claraparabricks/genomeworks/cudaextender/extender.hpp

```diff
@@ -0,0 +1,158 @@
+    /// \param[in] query_length Length of query sequence
+    /// \param[in] d_target Device pointer to encoded target sequence
+    /// \param[in] target_length Length of target sequence
+    /// \param[in] score_threshold Score threshold for filtering extended segments
+    /// \param[in] d_seed_pairs Device pointer to array of seed pairs mapping between
+    ///                         target and query sequences
+    /// \param[in] num_seed_pairs Length of d_seed_pairs array
+    /// \param[out] d_scored_segment_pairs Pointer to a pre-allocated device location for
+    ///                                    storing extension output
+    /// \param[out] d_num_scored_segment_pairs Pointer to pre-allocated device location for
+    ///                                        storing length of extension output
+    /// \return Status of the async extension launch
+    virtual StatusType extend_async(const int8_t* d_query, int32_t query_length,
+                                    const int8_t* d_target, int32_t target_length,
+                                    int32_t score_threshold, SeedPair* d_seed_pairs,
+                                    int32_t num_seed_pairs, ScoredSegmentPair* d_scored_segment_pairs,
+                                    int32_t* d_num_scored_segment_pairs) = 0;
+
+    /// \brief Waits for CUDA accelerated extension to finish
+    ///
+    /// To be used with the host pointer extend_async API.
+    /// Blocking call that waits for all the extensions scheduled
+    /// on the GPU to come to completion.
+    ///
+    /// \return Synchronization and memory copy status
+    virtual StatusType sync() = 0;
+
+    /// \brief Returns the computed scored segment pairs
+    ///
+    /// To be used with the host pointer extend_async API. sync() must
+    /// be called after launching async extension and before calling this function
+    /// or the returned results will not be valid.
+    ///
+    /// \return Vector of Scored Segment Pairs that have a score >= score_threshold
+    virtual const std::vector<ScoredSegmentPair>& get_scored_segment_pairs() const = 0;
+
+    /// \brief Reset Extender object and free device/host memory
+    virtual void reset() = 0;
+};
+
+/// \brief Create Extender object
+///
+/// \param h_score_mat Host pointer to scoring matrix for use during extension
+/// \param score_mat_dim Dimension of the scoring matrix
+/// \param xdrop_threshold Threshold for performing X-Drop
+/// \param no_entropy Flag indicating whether to use entropy during extension
+/// \param stream CUDA Stream to be used with extension
+/// \param device_id GPU to be used for extension
+/// \param allocator DeviceAllocator to be used for allocating/freeing memory
+/// \param type Type of extension to be performed
+/// \return Unique pointer to Extender object.
+std::unique_ptr<Extender> create_extender(const int32_t* h_score_mat,
+                                          const int32_t score_mat_dim,
```

Comment:
Please remove the `const` from the parameters that get passed by value. (Explanation see earlier comment.)

Reply:
Done.

Reply:
That went a bit too far. :)
`h_score_mat` should still be `const int32_t* h_score_mat`.
The function is getting a "pointer to constant memory" passed by value. For the caller of the function, it is important to know that we will not modify the memory this pointer points to.

In contrast, the pointer analog to the integer with const parameter (`const int32_t`) would be
`const int32_t* const h_score_mat`, i.e. a constant pointer to constant memory. This means we cannot reassign the pointer `h_score_mat` to point to other memory within our function, but this reassignment would have no effect for the caller.

Reply:
My bad. Changed the pointers to const type*